File: leetcode2/second/0671. Second Minimum Node In a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findSecondMinimumValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return -1
        return self.findbigVal(root,root.val)
    
    def findbigVal(self,root,target):
        if not root:
            return -1
        if root.val!=target:
            return root.val
        if root.val==target:
            left=self.findbigVal(root.left,target)
            right=self.findbigVal(root.right,target)
            if left==-1 and right==-1:
                return -1
            elif left!=-1 and right==-1:
                return left
            elif left==-1 and right!=-1:
                return right
            else:
                return min(left,right)
                

File: archives/C#/0671. Second Minimum Node In a Binary Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int FindSecondMinimumValue(TreeNode root) {
        if(root==null){
            return -1;
        }
        int target=root.val;
        int result=int.MaxValue;
        Stack<TreeNode> rootStack=new Stack<TreeNode>();
        rootStack.Push(root);
        while(rootStack.Count()!=0){
            TreeNode curRoot=rootStack.Pop();
            if(curRoot.val!=target){
                result=Math.Min(result,curRoot.val);
            }
            if(curRoot.left!=null){
                rootStack.Push(curRoot.left);
                rootStack.Push(curRoot.right);
            }
        }
        return result==int.MaxValue?-1:result;
    }
}
